### PR TITLE
allow names to change capitalisation in patch requests #265

### LINKED ISSUE #265: Allow names to change capitalisation
The current `_is_duplicate` checks for catalogue categories, manufacturers and systems don't take into account capitalisation changes. e.g. editing `test` to `Test` isn't allowed. We effectively also need to check that the returned document id is different to the one being edited. The check for create, move and copy should still error if the name is similar.

### PR DESCRIPTION
**Pull request by joshuadkitenge**
## Description
see #265

## Testing instructions
- [ ] Review code
- [ ] Check Actions build

## Agile board tracking
closes #265

### CODE REVIEW

File: test/e2e/test_catalogue_category.py [hunk 1]
```diff
@@ -744,6 +744,22 @@ def test_partial_update_catalogue_category_change_name(test_client):
     }
 
 
+def test_partial_update_catalogue_category_change_captialistion_of_name(test_client):
```
Comment (joelvdavies) [inline test/e2e/test_catalogue_category.py:747]:
```suggestion
def test_partial_update_catalogue_category_change_capitalisation_of_name(test_client):
```

File: test/e2e/test_system.py [hunk 2]
```diff
@@ -440,6 +440,21 @@ def test_partial_update_system_name(test_client):
...
+    Test updating a captialistion of the System's name
+    """
+    _test_partial_update_system(
+        test_client,
+        {
+            "name": "SyStEm A",
+        },
+        {
+            "name": "SyStEm A",
+        },
```
Comment (joelvdavies) [inline test/e2e/test_system.py:454]:
```suggestion
```
This part is only needed if code is changed.

File: test/unit/repositories/test_catalogue_category.py [hunk 3]
```diff
@@ -1086,6 +1086,97 @@ def test_update_duplicate_name_within_new_parent(test_helpers, database_mock, ca
     assert str(exc.value) == "Duplicate catalogue category found within the parent catalogue category"
 
 
+def test_update_change_captialistion_of_name_within_parent(test_helpers, database_mock, catalogue_category_repository):
+    """
+    Test updating a catalogue category when the code is the same and the captialisation of the name has changed.
```
Comment (joelvdavies) [inline test/unit/repositories/test_catalogue_category.py:1091]:
```suggestion
def test_update_change_capitalisation_of_name_within_parent(test_helpers, database_mock, catalogue_category_repository):
    """
    Test updating a catalogue category when the code is the same and the capitalisation of the name has changed.
```

File: test/unit/repositories/test_catalogue_category.py [hunk 4]
```diff
@@ -1086,6 +1086,97 @@ def test_update_duplicate_name_within_new_parent(test_helpers, database_mock, ca
...
+        {
+            **CATALOGUE_CATEGORY_INFO,
+            **MOCK_CREATED_MODIFIED_TIME,
+            "_id": CustomObjectId(catalogue_category.id),
+            "is_leaf": catalogue_category.is_leaf,
+            "parent_id": catalogue_category.parent_id,
+            "catalogue_item_properties": catalogue_category.catalogue_item_properties,
+        },
+    )
+    # Mock `find_one` to return no duplicate catalogue categories found
```
Comment (joelvdavies) [inline test/unit/repositories/test_catalogue_category.py:1122]:
```suggestion
    # Mock `find_one` to return a duplicate catalogue category but with the same id as the one being updated
```

File: test/unit/repositories/test_manufacturer.py [hunk 5]
```diff
@@ -247,6 +247,89 @@ def test_get_with_nonexistent_id(test_helpers, database_mock, manufacturer_repos
...
+                "town": "Oxford",
+                "county": "Oxfordshire",
+                "postcode": "OX1 2AB",
+                "country": "United Kingdom",
+            },
+            "telephone": "0932348348",
+        },
+    )
+
+    # Mock `find_one` to return no duplicate manufacturers found
```
Comment (joelvdavies) [inline test/unit/repositories/test_manufacturer.py:291]:
```suggestion
    # Mock `find_one` to return a duplicate manufacturer but with the same id as the one being updated
```

File: test/unit/repositories/test_system.py [hunk 6]
```diff
@@ -632,6 +632,59 @@ def test_update_duplicate_name_within_parent(test_helpers, database_mock, system
...
+    system = SystemOut(id=str(ObjectId()), **SYSTEM_A_INFO, **MOCK_CREATED_MODIFIED_TIME)
+    session = MagicMock()
+
+    # Mock `find_one` to return the stored System document
+    test_helpers.mock_find_one(
+        database_mock.systems,
+        system.model_dump(),
+    )
+
+    # Mock `find_one` to return duplicate system found in parent system
```
Comment (joelvdavies) [inline test/unit/repositories/test_system.py:651]:
```suggestion
    # Mock `find_one` to return a duplicate system but with the same id as the one being updated
```

**Review by joelvdavies — CHANGES_REQUESTED** [review]
Minor comments worked well with https://github.com/ral-facilities/inventory-management-system/pull/554.

File: inventory_management_system_api/repositories/catalogue_category.py [hunk 7]
```diff
@@ -201,25 +203,32 @@ def list(self, parent_id: Optional[str], session: ClientSession = None) -> List[
...
+        catalogue_category_id: CustomObjectId = None,
+        session: ClientSession = None,
     ) -> bool:
         """
         Check if a catalogue category with the same code already exists within the parent category.
 
         :param parent_id: The ID of the parent catalogue category which can also be `None`.
         :param code: The code of the catalogue category to check for duplicates.
+        :param catalogue_category_id: The ID of the catalogue category to check if the duplicate
+         catalogue category found is itself.
```
Comment (VKTB) [inline inventory_management_system_api/repositories/catalogue_category.py:218]: Could you please indent this line for better readability?

File: inventory_management_system_api/repositories/catalogue_category.py [hunk 8]
```diff
@@ -201,25 +203,32 @@ def list(self, parent_id: Optional[str], session: ClientSession = None) -> List[
...
         :return: `True` if a duplicate catalogue category code is found, `False` otherwise.
         """
         logger.info("Checking if catalogue category with code '%s' already exists within the parent category", code)
         if parent_id:
             parent_id = CustomObjectId(parent_id)
 
-        return (
-            self._catalogue_categories_collection.find_one({"parent_id": parent_id, "code": code}, session=session)
-            is not None
+        catalogue_category_with_code = self._catalogue_categories_collection.find_one(
```
Comment (VKTB) [inline inventory_management_system_api/repositories/catalogue_category.py:226]: Calling this `catalogue_category` should be just fine I think
```suggestion
        catalogue_categorye = self._catalogue_categories_collection.find_one(
```

File: inventory_management_system_api/repositories/catalogue_category.py [hunk 9]
```diff
@@ -201,25 +203,32 @@ def list(self, parent_id: Optional[str], session: ClientSession = None) -> List[
...
+            {"parent_id": parent_id, "code": code}, session=session
```
Comment (VKTB) [inline inventory_management_system_api/repositories/catalogue_category.py:227]: Here you can also do something like `{"parent_id": parent_id, "code": code, "_id": {"$ne": catalogue_category_id}}, session=session` but I will leave it to you to decide if you want to change it as it works just fine with how it is at the moment.

File: inventory_management_system_api/repositories/manufacturer.py [hunk 10]
```diff
@@ -133,16 +133,20 @@ def delete(self, manufacturer_id: str, session: ClientSession = None) -> None:
...
         Check if manufacturer with the same name already exists in the manufacturer collection
 
         :param code: The code of the manufacturer to check for duplicates.
+        :param manufacturer_id: The ID of the manufacturer to check if the duplicate manufacturer found is itself.
         :param session: PyMongo ClientSession to use for database operations
         :return `True` if duplicate manufacturer, `False` otherwise
         """
         logger.info("Checking if manufacturer with code '%s' already exists", code)
-        return self._manufacturers_collection.find_one({"code": code}, session=session) is not None
+        manufacturer_with_code = self._manufacturers_collection.find_one({"code": code}, session=session)
```
Comment (VKTB) [inline inventory_management_system_api/repositories/manufacturer.py:148]: Calling this `manufacturer` should be just fine I think
```suggestion
        manufacturer = self._manufacturers_collection.find_one({"code": code}, session=session)
```

File: inventory_management_system_api/repositories/system.py [hunk 11]
```diff
@@ -177,20 +177,24 @@ def delete(self, system_id: str, session: ClientSession = None) -> None:
...
+        :param system_id: The ID of the system to check if the duplicate system found is itself.
         :param session: PyMongo ClientSession to use for database operations
         :return: `True` if a duplicate System code is found under the given parent, `False` otherwise
         """
         logger.info("Checking if System with code '%s' already exists within the parent System", code)
         if parent_id:
             parent_id = CustomObjectId(parent_id)
 
-        return self._systems_collection.find_one({"parent_id": parent_id, "code": code}, session=session) is not None
+        system_with_code = self._systems_collection.find_one({"parent_id": parent_id, "code": code}, session=session)
```
Comment (VKTB) [inline inventory_management_system_api/repositories/system.py:196]: Calling this `system` should be just fine I think
```suggestion
        system = self._systems_collection.find_one({"parent_id": parent_id, "code": code}, session=session)
```

File: test/e2e/test_manufacturer.py [hunk 12]
```diff
@@ -194,6 +194,18 @@ def test_partial_address_update(test_client):
     }
 
 
+def test_update_captialistion_of_name(test_client):
```
Comment (VKTB) [inline test/e2e/test_manufacturer.py:197]:
```suggestion
def test_update_capitalisation_of_name(test_client):
```

For #90 - The tests in this module should be checked and renamed if necessary. The update tests should be renamed so they start with `test_partial_update..` to keep them consistent with the e2e tests in the other modules.

File: test/e2e/test_manufacturer.py [hunk 13]
```diff
@@ -194,6 +194,18 @@ def test_partial_address_update(test_client):
...
+    """Test updating a manufacturer when the captialistion of the name is different"""
```
Comment (VKTB) [inline test/e2e/test_manufacturer.py:198]:
```suggestion
    """Test updating a manufacturer when the capitalisation of the name is different"""
```

File: test/e2e/test_catalogue_category.py [hunk 14]
```diff
@@ -744,6 +744,22 @@ def test_partial_update_catalogue_category_change_name(test_client):
     }
 
 
+def test_partial_update_catalogue_category_change_capitalisation_of_name(test_client):
+    """
+    Test changing the captialistion of the name of a catalogue category.
```
Comment (VKTB) [inline test/e2e/test_catalogue_category.py:749]:
```suggestion
    Test changing the capitalisation of the name of a catalogue category.
```

File: test/e2e/test_system.py [hunk 15]
```diff
@@ -440,6 +440,18 @@ def test_partial_update_system_name(test_client):
     _test_partial_update_system(test_client, {"name": "Updated name"}, {"code": "updated-name"})
 
 
+def test_partial_update_captialistion_of_system_name(test_client):
```
Comment (VKTB) [inline test/e2e/test_system.py:443]:
```suggestion
def test_partial_update_capitalisation_of_system_name(test_client):
```

File: test/e2e/test_system.py [hunk 16]
```diff
@@ -440,6 +440,18 @@ def test_partial_update_system_name(test_client):
...
+    """
+    Test updating a captialistion of the System's name
```
Comment (VKTB) [inline test/e2e/test_system.py:445]:
```suggestion
    Test updating a capitalisation of the System's name
```

File: test/unit/repositories/test_manufacturer.py [hunk 17]
```diff
@@ -247,6 +247,89 @@ def test_get_with_nonexistent_id(test_helpers, database_mock, manufacturer_repos
     database_mock.manufacturers.find_one.assert_called_once_with({"_id": CustomObjectId(manufacturer_id)}, session=None)
 
 
+def test_update_change_captialistion_of_name(test_helpers, database_mock, manufacturer_repository):
```
Comment (VKTB) [inline test/unit/repositories/test_manufacturer.py:250]:
```suggestion
def test_update_change_capitalisation_of_name(test_helpers, database_mock, manufacturer_repository):
```

File: test/unit/repositories/test_manufacturer.py [hunk 18]
```diff
@@ -247,6 +247,89 @@ def test_get_with_nonexistent_id(test_helpers, database_mock, manufacturer_repos
...
+    """Test updating a manufacturer when the code is the same and the captialisation of the name has changed."""
```
Comment (VKTB) [inline test/unit/repositories/test_manufacturer.py:251]:
```suggestion
    """Test updating a manufacturer when the code is the same and the capitalisation of the name has changed."""
```

File: test/unit/repositories/test_system.py [hunk 19]
```diff
@@ -632,6 +632,59 @@ def test_update_duplicate_name_within_parent(test_helpers, database_mock, system
     database_mock.systems.update_one.assert_not_called()
 
 
+@patch("inventory_management_system_api.repositories.system.utils")
+def test_update_change_captialistion_of_name(utils_mock, test_helpers, database_mock, system_repository):
```
Comment (VKTB) [inline test/unit/repositories/test_system.py:636]:
```suggestion
def test_update_change_capitalisation_of_name(utils_mock, test_helpers, database_mock, system_repository):
```

File: test/unit/repositories/test_system.py [hunk 20]
```diff
@@ -632,6 +632,59 @@ def test_update_duplicate_name_within_parent(test_helpers, database_mock, system
...
+    """
+    Test updating a System when the code is the same and the captialisation of the name has changed.
```
Comment (VKTB) [inline test/unit/repositories/test_system.py:638]:
```suggestion
    Test updating a System when the code is the same and the capitalisation of the name has changed.
```

**Review by VKTB — CHANGES_REQUESTED** [review]

File: test/e2e/test_manufacturer.py (hunk 12 above)
Comment (joelvdavies) [reply test/e2e/test_manufacturer.py]: There are a few others that haven't been renamed.

Comment (VKTB) [reply test/e2e/test_manufacturer.py]: I was suggesting that they get addressed in #90.

Comment (joelvdavies) [reply test/e2e/test_manufacturer.py]: Oh right yes I see sorry.

**Review by joelvdavies — APPROVED** [review]

**Review by VKTB — APPROVED** [review]

**Pull request merged**